src/Components/Server/src/Circuits/CircuitHost.cs

```diff
@@ -136,6 +139,88 @@ internal void SendPendingBatches()
             var _ = Renderer.Dispatcher.InvokeAsync(() => Renderer.ProcessBufferedRenderBatches());
         }
 
+        internal async Task EndInvokeDotNetFromJS(string arguments)
+        {
+            try
+            {
+                AssertInitialized();
+
+                await Renderer.Dispatcher.InvokeAsync(() =>
+                {
+                    SetCurrentCircuitHost(this);
+                    var document = JsonDocument.Parse(arguments);
+                    if (document.RootElement.ValueKind != JsonValueKind.Array)
+                    {
+                        // Log error and return
+                        return;
+                    }
+                    var length = document.RootElement.GetArrayLength();
+                    if (length != 3)
+                    {
+                        // Log error and return
+                        return;
+                    }
+                    long asyncHandle = 0;
+                    bool succeeded = false;
+                    JSAsyncCallResult result = null;
+                    var i = 0;
+                    foreach (var element in document.RootElement.EnumerateArray())
+                    {
+                        if (i == 0)
+                        {
+                            asyncHandle = element.GetInt64();
+                        }
+                        if (i == 1)
+                        {
+                            succeeded = element.GetBoolean();
+                        }
+
+                        if (i == 2 && element.ValueKind != JsonValueKind.Null)
+                        {
+                            result = (JSAsyncCallResult)typeof(JSAsyncCallResult).GetConstructor(
+                                BindingFlags.NonPublic | BindingFlags.Instance,
+                                binder: null,
+                                new Type[] { typeof(JsonDocument), typeof(JsonElement) },
+                                null).Invoke(new object[] { document, element });
+                        }
+
+                        i++;
+                    }
+
+                    DotNetDispatcher.EndInvoke(asyncHandle, succeeded, result);
```

Comment:
One positive change that could come from this is that we could limit the use of the magic-static here to Mono/Wasm. Right now this is backed by an async local because that's the only reasonable way for this to work over pure JS interop.

Reply:
This makes the Mono implementation simpler as well, because in the Mono world there will be one at a time.

Reply:
We could have dotnet dispatcher not be a static class and have an associated JS runtime and then have a static for mono/wasm or simply return a dotnet object ref to a singleton instance and use that on the client-side to tunnel js interop.

Reply:
We got rid of delivering the callbacks through interop in favor of an explicit contract on both environments.